### PR TITLE
feat: agent podman support

### PR DESCRIPTION
**Pull request by phillip-gaisford**
## Description
Analogous to the singularity/apptainer support, here's something that works with podman.

I have made some minor changes to singularity.go to enable code sharing with the new podman.go.

## Test Plan
TBD

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: agent/pkg/podman/podman.go [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+	return &PodmanClient{
+		SingularityClient: client,
+	}, nil
+}
+
+// getPullCommand returns the command and arguments to perform the container image pull.
+func (s *PodmanClient) getPullCommand(req docker.PullImage, image string) (string, []string) {
+	args := []string{"pull"}
+	// if req.ForcePull {
+	// 	args = append(args, "--force") // Use 'podman image rm'?
```
Comment (jerryharrow) [inline agent/pkg/podman/podman.go:69]: Doesn't seem to be a pull option for this, but a...

```
podman run
      --pull string                              Pull image before creating ("always"|"missing"|"never") (default "missing")
```

File: agent/pkg/podman/podman.go [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+	tmpdir, err := os.MkdirTemp(agentTmp, fmt.Sprintf("*-%s", id))
+	if err != nil {
+		return nil, fmt.Errorf("making tmp dir for archives: %w", err)
+	}
+
+	var args []string
+	args = append(args, "run")
+	args = append(args, "--rm")
+	// args = append(args, "--writable-tmpfs")
```
Comment (jerryharrow) [inline agent/pkg/podman/podman.go:119]: Yup, probably not needed for podman.

File: agent/pkg/podman/podman.go [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+		}
+		req.HostConfig.NetworkMode = hostNetworking
+		fallthrough
+	case req.HostConfig.NetworkMode == hostNetworking:
+		args = append(args, "--network=host")
+	default:
+		return nil, fmt.Errorf("unsupported network mode %s", req.HostConfig.NetworkMode)
+	}
+
+	archivesPath := filepath.Join(tmpdir, archivesName)
```
Comment (jerryharrow) [inline agent/pkg/podman/podman.go:174]: Seems like the following section (archived expansion) should be a shared method of some time.  Only the CLI syntax is different, right?

File: agent/pkg/podman/podman.go (hunk 1 above)
Comment (phillip-gaisford) [reply agent/pkg/podman/podman.go:69]: If we do a `podman pull`, is there any benefit to using `--pull always` on the subsequent `podman run`? It's hard to say from the docs I have read.

Comment (jerryharrow) [reply agent/pkg/podman/podman.go:69]: I was commenting on the commented out --force which doesn't apply to podman.   

If we just do explicit pulls for podman, then it is doing a force.

So perhaps nothing is needed.

Comment (phillip-gaisford) [reply agent/pkg/podman/podman.go:69]: That's what I was thinking too.

**Comment by stoksc** [comment]
we should not use embedding as inheritance. if there is duplicated logic, pull out helper functions and put them in a `cruntimes` package that is shared or something.

**Comment by phillip-gaisford** [comment]
Superseded by: https://github.com/determined-ai/determined/pull/6657

**Pull request closed without merging**